File: src/main/java/com/peasenet/mods/misc/ModGuiTextOverlay.java
/*
 * Copyright (c) 2022. Gavin Pease and contributors.
 *
 * Permission is hereby granted, free of charge, to any person obtaining a copy of this software and
 * associated documentation files (the "Software"), to deal in the Software without restriction, including
 * without limitation the rights to use, copy, modify, merge, publish, distribute, sublicense, and/or sell copies
 *  of the Software, and to permit persons to whom the Software is furnished to do so, subject to the
 *  following conditions:
 *
 * The above copyright notice and this permission notice shall be included in all copies or substantial
 * portions of the Software.
 *
 * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR IMPLIED,
 * INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY, FITNESS FOR A PARTICULAR
 * PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT HOLDERS BE
 * LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT
 * OR OTHERWISE, ARISING FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR
 * OTHER DEALINGS IN THE SOFTWARE.
 */

package com.peasenet.mods.misc;

import com.peasenet.main.GavinsMod;
import com.peasenet.main.GavinsModClient;
import com.peasenet.main.Settings;
import com.peasenet.mods.Mod;
import com.peasenet.mods.Type;
import com.peasenet.settings.ToggleSetting;
import com.peasenet.util.RenderUtils;
import com.peasenet.util.math.BoxD;
import com.peasenet.util.math.PointD;
import net.minecraft.client.util.math.MatrixStack;
import net.minecraft.text.Text;

import java.util.Comparator;
import java.util.concurrent.atomic.AtomicInteger;

/**
 * @author gt3ch1
 * @version 6/28/2022
 * A mod that shows the currently active mods in the top left screen.
 */
public class ModGuiTextOverlay extends Mod {
    public ModGuiTextOverlay() {
        super(Type.MOD_GUI_TEXT_OVERLAY);

        //NOTE: This isn't really the best place for this, but it works for now. this is for chat message toggles.
        ToggleSetting chatMessage = new ToggleSetting("chatMessage", "gavinsmod.settings.misc.messages");
        chatMessage.setValue(Settings.getBool("chatMessage"));
        addSetting(chatMessage);
    }

    @Override
    public void onRenderInGameHud(MatrixStack matrixStack, float delta) {
        if (GavinsMod.isEnabled(Type.MOD_GUI) || GavinsMod.isEnabled(Type.SETTINGS))
            return;
        drawTextOverlay(matrixStack);
    }


    /**
     * Draws the GUI text overlay if enabled, and if the main menu is not open.
     *
     * @param matrixStack - The matrix stack to use.
     */
    private void drawTextOverlay(MatrixStack matrixStack) {
        var textRenderer = GavinsModClient.getMinecraftClient().getTextRenderer();
        var startingPoint = new PointD(0, 0);
        int currX = (int) (startingPoint.x() + 2);
        AtomicInteger currY = new AtomicInteger((int) (startingPoint.y() + 2));
        if (GavinsMod.isEnabled(Type.MOD_GUI) || !GavinsMod.isEnabled(Type.MOD_GUI_TEXT_OVERLAY) || GavinsMod.isEnabled(Type.SETTINGS))
            return;
        // only get active mods, and mods that are not gui type.
        var mods = GavinsMod.getModsForTextOverlay();
        var modsCount = (int) GavinsMod.getModsForTextOverlay().count();
        if (modsCount == 0)
            return;
        // get the mod with the longest name.
        var longestModName = mods.max(Comparator.comparingInt(mod -> mod.getName().length())).get().getName().length();
        var box = new BoxD(startingPoint, longestModName * 6 + 6, modsCount * 12);
        RenderUtils.drawBox((Settings.getColor("backgroundColor")).getAsFloatArray(), box, matrixStack);
        mods = GavinsMod.getModsForTextOverlay();
        AtomicInteger modCounter = new AtomicInteger();
        mods.forEach(mod -> {
            textRenderer.draw(matrixStack, Text.translatable(mod.getTranslationKey()), currX, currY.get(), (Settings.getColor("foregroundColor")).getAsInt());
            if (modsCount > 1 && modCounter.get() < modsCount - 1) {
                RenderUtils.drawSingleLine((Settings.getColor("foregroundColor")).getAsFloatArray(), currX - 1, currY.get() + 9, longestModName * 6 + 5, currY.get() + 9, matrixStack);
            }
            currY.addAndGet(12);
            modCounter.getAndIncrement();
        });
    }
}

File: src/main/kotlin/com/peasenet/mods/misc/ModGuiTextOverlay.kt
/*
 * MIT License
 *
 * Copyright (c) 2022-2025, Gavin C. Pease
 *
 * Permission is hereby granted, free of charge, to any person obtaining a copy
 * of this software and associated documentation files (the "Software"), to deal
 * in the Software without restriction, including without limitation the rights
 * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
 * copies of the Software, and to permit persons to whom the Software is
 * furnished to do so, subject to the following conditions:
 *
 * The above copyright notice and this permission notice shall be included in all
 * copies or substantial portions of the Software.
 *
 * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
 * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
 * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
 * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
 * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
 * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
 * SOFTWARE.
 */
package com.peasenet.mods.misc

import com.mojang.blaze3d.systems.RenderSystem
import com.peasenet.gavui.math.BoxF
import com.peasenet.gavui.math.PointF
import com.peasenet.gavui.util.GavUISettings
import com.peasenet.gavui.util.GuiUtil
import com.peasenet.main.GavinsMod
import com.peasenet.main.GavinsModClient
import com.peasenet.main.Mods
import com.peasenet.mods.Mod
import com.peasenet.mods.gui.GuiMod
import com.peasenet.util.RenderUtils
import com.peasenet.util.listeners.InGameHudRenderListener
import net.minecraft.client.gui.GuiGraphicsExtractor
import net.minecraft.client.resources.language.I18n
import net.minecraft.network.chat.Component
import net.minecraft.world.phys.Vec3

/**
 * @author GT3CH1
 * @version 01-15-2025
 * A mod that shows the currently active mods in the top left screen.
 */
class ModGuiTextOverlay : MiscMod(
    "gavinsmod.mod.misc.textoverlay",
    "textoverlay",
), InGameHudRenderListener {

    override fun onEnable() {
        super.onEnable()
        em.subscribe(InGameHudRenderListener::class.java, this)
    }

    override fun onDisable() {
        super.onDisable()
        em.unsubscribe(InGameHudRenderListener::class.java, this)
    }

    override fun onRenderInGameHud(drawContext: GuiGraphicsExtractor, delta: Float, forceRender: Boolean) {
        modList = Mods.mods.filter { mod: Mod -> mod !is GuiMod && mod !is ModGuiTextOverlay && mod.isActive }
        if (GavinsMod.isEnabled("gui") || GavinsMod.isEnabled("settings") || modList.isEmpty() || forceRender) return
        drawTextOverlay(drawContext)
    }

    /**
     * Draws the GUI text overlay if enabled, and if the main menu is not open.
     *
     * @param drawContext- The drawContext to use
     */
    private fun drawTextOverlay(drawContext: GuiGraphicsExtractor) {
        val matrixStack = drawContext.pose()
        val textRenderer = GavinsModClient.minecraftClient.textRenderer
        var startingPoint = PointF(0f, 0f)
        val modsCount = modList.count()
        var longestModName = 0
        for (mod in modList) {
            longestModName = textRenderer.width(I18n.get(mod.translationKey)).coerceAtLeast(longestModName)
        }
        val box = BoxF(startingPoint, (longestModName + 4).toFloat(), modsCount * 12f)
        matrixStack.pushMatrix()
        val color = GavUISettings.getColor("gui.color.background")
        val alpha = GavUISettings.getFloat("gui.alpha")
        GuiUtil.fill(
            box,
            drawContext, color.withAlpha(alpha))
        drawContext.guiRenderState.up()
        drawContext.enableScissor(0,0, GavinsModClient.minecraftClient.window.width, GavinsModClient.minecraftClient.window.height)
        for ((index, mod) in modList.withIndex()) {
             drawContext.text(
                textRenderer,
                Component.translatable(mod.translationKey),
                startingPoint.x.toInt() + 2,
                startingPoint.y.toInt() + 2,
                GavUISettings.getColor("gui.color.foreground").asInt,
                false
            )
            if (modsCount > 1 && index < modsCount - 1) {
                val p1 = Vec3(0.0, startingPoint.y + 11.0, 0.0)
                val p2 = Vec3(longestModName + 3.0, startingPoint.y + 11.5, 0.0)
                drawContext.horizontalLine(p1.x.toInt(), p2.x.toInt(), p1.y.toInt(), GavUISettings.getColor("gui.color.border").asInt)
            }
            startingPoint = startingPoint.add(0.0f, 12f)
        }
        drawContext.disableScissor()
        matrixStack.popMatrix()


    }

    companion object {
        lateinit var modList: List<Mod>
    }
}
